nixos/modules/services/taler/module.nix

```diff
@@ -0,0 +1,108 @@
+{
+  lib,
+  pkgs,
+  config,
+  ...
+}:
+
+let
+  cfg = config.services.taler;
+  settingsFormat = pkgs.formats.ini { };
+in
+
+{
+  # TODO turn this into a generic taler-like service thingy?
+  options.services.taler = {
+    enable = lib.mkEnableOption "the GNU Taler system" // lib.mkOption { internal = true; };
+    includes = lib.mkOption {
+      type = lib.types.listOf lib.types.path;
+      default = [ ];
+      description = ''
+        Files to include into the config file using Taler's `@inline@` directive.
+
+        This allows including arbitrary INI files, including imperatively managed ones.
+      '';
+    };
+    configFile = lib.mkOption {
+      internal = true;
+      default =
+        let
+          includes = pkgs.writers.writeText "includes.conf" (
+            lib.concatStringsSep "\n" (map (include: "@inline@ ${include}") cfg.includes)
+          );
+          generatedConfig = settingsFormat.generate "generated-taler.conf" cfg.settings;
+        in
+        pkgs.runCommand "taler.conf" { } ''
+          cat ${includes} > $out
+          echo >> $out
+          echo >> $out
+          cat ${generatedConfig} >> $out
+        '';
+    };
+    settings = lib.mkOption {
+      description = ''
+        Global configuration options for the taler config file.
+
+        For a list of all possible options, please see the man page [`taler.conf(5)`](https://docs.taler.net/manpages/taler.conf.5.html)
+      '';
+      type = lib.types.submodule {
+        freeformType = settingsFormat.type;
+        options = {
+          # Should these be in here?
+          # Should these even be configurable?
+          PATHS = {
+            TALER_DATA_HOME = lib.mkOption {
+              type = lib.types.str;
+              internal = true;
+              default = "\${STATE_DIRECTORY}/";
+            };
+            TALER_CACHE_HOME = lib.mkOption {
+              type = lib.types.str;
+              internal = true;
+              default = "\${CACHE_DIRECTORY}/";
+            };
+            TALER_RUNTIME_DIR = lib.mkOption {
+              type = lib.types.str;
+              internal = true;
+              default = cfg.runtimeDir;
+            };
+          };
```

Comment:
Maybe these could be set with these default values in `config` instead? It might make the code less awkward. Then people can override them on their own.

```nix
config = lib.mkIf cfg.enable {
  ...
  services.taler.settings.PATHS = {
    TALER_DATA_HOME = lib.mkDefault "\${STATE_DIRECTORY}/";
    TALER_CACHE_HOME = lib.mkDefault "\${CACHE_DIRECTORY}/";
    TALER_RUNTIME_DIR = cfg.runtimeDir;
  };
};
```

Reply:
I don't know if normal users should change these options as functionality can break if they're misconfigured.

Reply:
I'm a bit wary of setting things so that the upstream users cannot modify them whatsoever. People tend to have weird usecases for things that I hadn't thought about before. Like for example, someone might want to store the `DATA_HOME` on a different drive and for some reason can't resort to bindmounts, or want to store their cache in an in-memory disk.

If you remove the `lib.mkDefault`, people would have to use `mkForce` as if to say "I know what I'm doing". How does that sound?